# Python Language Basics, IPython, and Jupyter Notebooks

## 2.1 The Python interpreter

Run script stored in file from Jupyter notebook.

You can call Python interpreter by calling `python` in terminal. You run scripts using it. Exit it by calling `exit()`. 

You can also call also use IPython console - an enhanced Python interpreter - by calling `ipython`.

Run scripts from IPython console by clalling `%run hello_world.py`.

The last alternative is using Jupyter notebook by calling `jupyter notebook` (http://localhost:8888/).

## 2.2 IPython Basics

### Running the IPython Shell

Use pretty print for clarity.

In [2]:
import numpy as np

In [4]:
data = {i : np.random.randn() for i in range(7)}
data

{0: 0.17842767476662136,
 1: 0.5460357038062909,
 2: -0.2520211661018981,
 3: -2.4826966714240557,
 4: 1.3803935975974275,
 5: 0.43632834306677587,
 6: 0.2918482206018506}

In [5]:
print(data)

{0: 0.17842767476662136, 1: 0.5460357038062909, 2: -0.2520211661018981, 3: -2.4826966714240557, 4: 1.3803935975974275, 5: 0.43632834306677587, 6: 0.2918482206018506}


### Running the Jupyter Notebook

One of the major components of the Jupyter project is the notebook, a type of interactive
document for code, text (with or without markup), data visualizations, and other
output. 

The Jupyter notebook interacts with kernels, which are implementations of the Jupyter interactive computing protocol in any number of programming languages.
Python’s Jupyter kernel uses the IPython system for its underlying behavior.

Start Jupyter using `jupyter notebook` in the terminal.

### Tab Completion
On the surface, the IPython shell looks like a cosmetically different version of the
standard terminal Python interpreter (invoked with python). One of the major
improvements over the standard Python shell is tab completion, found in many IDEs
or other interactive computing analysis environments. While entering expressions in
the shell, pressing the Tab key will search the namespace for any variables (objects,
functions, etc.) matching the characters you have typed so far.


In [6]:
an_apple = 27
an_example = 42

Try `an_<tab>` here for completion

In [ ]:
an_

You can also complete
methods and attributes on any object after typing a period.

In [7]:
b = [1, 2, 3]

Use `b.<tab>` here to get object methods.

In [ ]:
b.

The same goes for modules.

In [8]:
import datetime

Use `datetime.<tab>`.

In [ ]:
datetime.

Note that IPython by default hides methods and attributes starting
with underscores, such as magic methods and internal “private”
methods and attributes, in order to avoid cluttering the display
(and confusing novice users!). These, too, can be tab-completed,
but you must first type an underscore to see them. If you prefer to
always see such methods in tab completion, you can change this
setting in the IPython configuration. See the IPython documentation
to find out how to do this.

When typing anything that looks
like a file path (even in a Python string), pressing the Tab key will complete anything
on your computer’s filesystem matching what you’ve typed.

Try typing `data<tab>`.

In [2]:
"data

README.md                          datasets/
ch_2_python_language_basics.ipynb  scripts/


Combined with the %run command this
functionality can save you many keystrokes.

Another area where tab completion saves time is in the completion of function keyword
arguments (and including the = sign!).

In [5]:
def func_with_keywords(abra=1, abbra=2, abbbra=3):
    return abra, abbra, abbbra

Type `func_with_keywords(ab<tab>`.

In [ ]:
func_with_keywords(ab

### Introspection

Using a question mark (?) before or after a variable will display some general information
about the object.

In [6]:
b = [1, 2, 3]

In [7]:
b?

In [8]:
print?

This is referred to as object introspection. If the object is a function or instance
method, the docstring, if defined, will also be shown.

In [9]:
def add_numbers(a, b):
    """
    Add two numbers together
    
    Returns
    -------
    the_sum : type of arguments
    """
    return a + b

In [10]:
add_numbers?

Using ?? will also show the function’s source code if possible.

In [11]:
add_numbers??

? has a final usage, which is for searching the IPython namespace in a manner similar
to the standard Unix or Windows command line. A number of characters combined
with the wildcard (*) will show all names matching the wildcard expression. For
example, we could get a list of all functions in the top-level NumPy namespace containing
load.

In [23]:
import numpy as np

In [24]:
np.*load*?

### The %run Command

You can run any file as a Python program inside the environment of your IPython
session using the %run command. Suppose you had the following simple script stored
in ipython_script_test.py.

In [26]:
%run scripts/ipython_script_test.py

The script is run in an empty namespace (with no imports or other variables defined)
so that the behavior should be identical to running the program on the command line
using python script.py. All of the variables (imports, functions, and globals)
defined in the file (up until an exception, if any, is raised) will then be accessible in
the IPython shell.

In [27]:
c

7.5

In [28]:
result

1.8

If a Python script expects command-line arguments (to be found in sys.argv), these
can be passed after the file path as though run on the command line.

Should you wish to give a script access to variables already defined
in the interactive IPython namespace, use %run -i instead of plain
%run.

In the Jupyter notebook, you may also use the related %load magic function, which
imports a script into a code cell.

In [ ]:
# %load scripts/ipython_script_test.py
def f(x, y, z):
	return (z + y) / z


a = 5
b = 6
c = 7.5

result = f(a, b, c)

### Interrupting running code

Pressing Ctrl-C while any code is running, whether a script through %run or a longrunning
command, will cause a KeyboardInterrupt to be raised. This will cause
nearly all Python programs to stop immediately except in certain unusual cases.

When a piece of Python code has called into some compiled extension
modules, pressing Ctrl-C will not always cause the program
execution to stop immediately. In such cases, you will have to
either wait until control is returned to the Python interpreter, or in
more dire circumstances, forcibly terminate the Python process.

### Executing code from the clipboard

If you are using the Jupyter notebook, you can copy and paste code into any code cell
and execute it. It is also possible to run code from the clipboard in the IPython shell.
Suppose you had the following code in some other application.

In [30]:
x = 5
y = 7
if x > 5:
    x += 1
    
    y = 8

The most foolproof methods are the %paste and %cpaste magic functions. %paste
takes whatever text is in the clipboard and executes it as a single block in the shell.

Copy the code above.

In [1]:
%paste

ERROR:root:Line magic function `%paste` not found.


%cpaste is similar, except that it gives you a special prompt for pasting code intom

In [2]:
%cpaste

ERROR:root:Line magic function `%cpaste` not found.


With the %cpaste block, you have the freedom to paste as much code as you like
before executing it. You might decide to use %cpaste in order to look at the pasted
code before executing it. If you accidentally paste the wrong code, you can break out
of the %cpaste prompt by pressing Ctrl-C.

### Terminal keyboar shortcuts

IPython has many keyboard shortcuts for navigating the prompt (which will be familiar
to users of the Emacs text editor or the Unix bash shell) and interacting with the
shell’s command history. 

```
Ctrl-P or up-arrow Search backward in command history for commands starting with currently entered text
Ctrl-N or down-arrow Search forward in command history for commands starting with currently entered text
Ctrl-R Readline-style reverse history search (partial matching)
Ctrl-Shift-V Paste text from clipboard
Ctrl-C Interrupt currently executing code
Ctrl-A Move cursor to beginning of line
Ctrl-E Move cursor to end of line
Ctrl-K Delete text from cursor until end of line
Ctrl-U Discard all text on current line
Ctrl-F Move cursor forward one character
Ctrl-B Move cursor back one character
Ctrl-L Clear screen
```


![alt text](resources/ch_2_1.png)

Note that Jupyter notebooks have a largely separate set of keyboard shortcuts for navigation
and editing. Since these shortcuts have evolved more rapidly than IPython’s, I
encourage you to explore the integrated help system in the Jupyter notebook’s menus.

### About magic commands

IPython’s special commands (which are not built into Python itself) are known as
“magic” commands. These are designed to facilitate common tasks and enable you to
easily control the behavior of the IPython system. A magic command is any command
prefixed by the percent symbol %. For example, you can check the execution
time of any Python statement, such as a matrix multiplication, using the %timeit
magic function (which will be discussed in more detail later).

In [4]:
import numpy as np

In [5]:
a = np.random.randn(100, 100)

In [6]:
%timeit np.dot(a, a)

The slowest run took 4648.50 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 34.3 µs per loop


Magic commands can be viewed as command-line programs to be run within the
IPython system. Many of them have additional “command-line” options, which can
all be viewed (as you might expect) using ?.

In [7]:
%debug?

Magic functions can be used by default without the percent sign, as long as no variable
is defined with the same name as the magic function in question. This feature is
called automagic and can be enabled or disabled with %automagic.

Some magic functions behave like Python functions and their output can be assigned
to a variable.

In [8]:
%pwd

'/Users/michalsznajder/Dropbox/github/wes_mckinney_python_for_data_analysis_2nd_ed'

In [9]:
foo = %pwd

In [10]:
foo

'/Users/michalsznajder/Dropbox/github/wes_mckinney_python_for_data_analysis_2nd_ed'

Since IPython’s documentation is accessible from within the system, I encourage you
to explore all of the special commands available by typing %quickref or %magic.

In [12]:
%magic

Table 2-2 highlights some of the most critical ones for being productive in interactive
computing and Python development in IPython.

```
%quickref Display the IPython Quick Reference Card
%magic Display detailed documentation for all of the available magic commands
%debug Enter the interactive debugger at the bottom of the last exception traceback
%hist Print command input (and optionally output) history
%pdb Automatically enter debugger after any exception
%paste Execute preformatted Python code from clipboard
%cpaste Open a special prompt for manually pasting Python code to be executed
%reset Delete all variables/names defined in interactive namespace
%page OBJECT Pretty-print the object and display it through a pager
%run script.py Run a Python script inside IPython
%prun statement Execute statement with cProfile and report the profiler output
%time statement Report the execution time of a single statement
%timeit statement Run a statement multiple times to compute an ensemble average execution time; useful for
timing code with very short execution time
%who, %who_ls, %whos Display variables defined in interactive namespace, with varying levels of information/
verbosity
%xdel variable Delete a variable and attempt to clear any references to the object in the IPython internals
```

### Matplotlib integration

One reason for IPython’s popularity in analytical computing is that it integrates well
with data visualization and other user interface libraries like matplotlib. 

The %matplotlib magic function configures its integration with the IPython
shell or Jupyter notebook. This is important, as otherwise plots you create will
either not appear (notebook) or take control of the session until closed (shell).

In the IPython shell, running %matplotlib sets up the integration so you can create
multiple plot windows without interfering with the console session.

In [13]:
%matplotlib

Using matplotlib backend: MacOSX


In Jupyter, the command is a little different.

In [14]:
%matplotlib inline

## 2.3 Python language basics

### Language semantics

The Python language design is distinguished by its emphasis on readability, simplicity,
and explicitness. Some people go so far as to liken it to “executable pseudocode.”

#### Indentation, not braces

Python uses whitespace (tabs or spaces) to structure code instead of using braces as in
many other languages like R, C++, Java, and Perl. Consider a for loop from a sorting
algorithm.

In [15]:
array = [1, 2, 3]
for x in array:
    if x < 2:
        print("a")
    else:
        print("b")

a
b
b


A colon denotes the start of an indented code block after which all of the code must
be indented by the same amount until the end of the block.

Love it or hate it, significant whitespace is a fact of life for Python programmers, and
in my experience it can make Python code more readable than other languages I’ve
used. While it may seem foreign at first, you will hopefully grow accustomed in time.

I strongly recommend using four spaces as your default indentation
and replacing tabs with four spaces. Many text editors have a setting
that will replace tab stops with spaces automatically (do this!).
Some people use tabs or a different number of spaces, with two
spaces not being terribly uncommon. By and large, four spaces is
the standard adopted by the vast majority of Python programmers,
so I recommend doing that in the absence of a compelling reason
otherwise.

Semicolons can be used, however, to separate multiple statements on a single
line.

In [16]:
a = 5; b = 6; c = 7

Putting multiple statements on one line is generally discouraged in Python as it often
makes code less readable.

#### Everything is an object

An important characteristic of the Python language is the consistency of its object
model. Every number, string, data structure, function, class, module, and so on exists
in the Python interpreter in its own “box,” which is referred to as a Python object.
Each object has an associated type (e.g., string or function) and internal data. In practice
this makes the language very flexible, as even functions can be treated like any
other object.

#### Comments

Any text preceded by the hash mark (pound sign) # is ignored by the Python interpreter.
This is often used to add comments to code. At times you may also want to
exclude certain blocks of code without deleting them. An easy solution is to comment
out the code.

In [17]:
# this is 
# not code

Comments can also occur after a line of executed code. While some programmers
prefer comments to be placed in the line preceding a particular line of code, this can
be useful at times.

In [18]:
print("Reached this line") # Simple status report

Reached this line


#### Function and object method calls

You call functions using parentheses and passing zero or more arguments, optionally
assigning the returned value to a variable.

In [21]:
def f(x, y, z):
    return x + y + z

f(2, 3, 4)

9

Almost every object in Python has attached functions, known as methods, that have
access to the object’s internal contents. You can call them using the following syntax.

In [22]:
t = [1, 2, 3]

In [23]:
t.append(4)

Functions can take both positional and keyword arguments.

In [24]:
def d(a, b, c, d=5, e="foo"):
    pass

d(1, 2, 3)

#### Variables and arguments passing

When assigning a variable (or name) in Python, you are creating a reference to the
object on the righthand side of the equals sign. In practical terms, consider a list of
integers.

In [25]:
a = [1, 2, 3]

Suppose we assign a to a new variable b.

In [26]:
b = a

In some languages, this assignment would cause the data [1, 2, 3] to be copied. In
Python, a and b actually now refer to the same object, the original list [1, 2, 3] (see
Figure 2-7 for a mockup). You can prove this to yourself by appending an element to
a and then examining b.

In [27]:
a.append(4)

In [28]:
b

[1, 2, 3, 4]

In [29]:
a

[1, 2, 3, 4]

Understanding the semantics of references in Python and when, how, and why data is
copied is especially critical when you are working with larger datasets in Python.

Assignment is also referred to as binding, as we are binding a name
to an object. Variable names that have been assigned may occasionally
be referred to as bound variables.

When you pass objects as arguments to a function, new local variables are created referencing
the original objects without any copying. If you bind a new object to a variable
inside a function, that change will not be reflected in the parent scope. It is
therefore possible to alter the internals of a mutable argument. Suppose we had the
following function.

In [30]:
def append_element(some_list, element):
    some_list.append(element)

In [32]:
data = [1, 2, 3]
append_element(data, 4)

In [33]:
data

[1, 2, 3, 4]

#### Dynamic references, strong types

In contrast with many compiled languages, such as Java and C++, object references in
Python have no type associated with them. There is no problem with the following.

In [34]:
a = 5
type(a)

int

In [35]:
a = "foo"
type(a)

str

Variables are names for objects within a particular namespace; the type information is
stored in the object itself. Some observers might hastily conclude that Python is not a
“typed language.” This is not true; consider this example.

In [36]:
'5' + 5

TypeError: must be str, not int

In some languages, such as Visual Basic, the string '5' might get implicitly converted
(or casted) to an integer, thus yielding 10. Yet in other languages, such as JavaScript,
the integer 5 might be casted to a string, yielding the concatenated string '55'. In this
regard Python is considered a strongly typed language, which means that every object
has a specific type (or class), and implicit conversions will occur only in certain obvious
circumstances, such as the following.

In [37]:
a = 4.5
b = 2

In [39]:
# String formatting, to be visited later
print("a is {0}, b is {1}".format(type(a), type(b)))

a is <class 'float'>, b is <class 'int'>


In [40]:
a / b

2.25

Knowing the type of an object is important, and it’s useful to be able to write functions
that can handle many different kinds of input. You can check that an object is an
instance of a particular type using the isinstance function.

In [41]:
a = 5

In [42]:
isinstance(a, int)

True

isinstance can accept a tuple of types if you want to check that an object’s type is
among those present in the tuple.

In [43]:
a = 5
b = 4.5

In [44]:
isinstance(a, (int, float))

True

In [45]:
isinstance(b, (int, float))

True

#### Attributes and methods

Objects in Python typically have both attributes (other Python objects stored “inside”
the object) and methods (functions associated with an object that can have access to
the object’s internal data). Both of them are accessed via the syntax
obj.attribute_name.

In [46]:
a = "foo"

Use `a.<tab>`.

In [ ]:
a.

Attributes and methods can also be accessed by name via the getattr function.

In [47]:
getattr(a, "split")

<function str.split>

In other languages, accessing objects by name is often referred to as “reflection.”
While we will not extensively use the functions getattr and related functions
hasattr and setattr in this book, they can be used very effectively to write generic,
reusable code.

#### Duck typing

Often you may not care about the type of an object but rather only whether it has
certain methods or behavior. This is sometimes called “duck typing,” after the saying
“If it walks like a duck and quacks like a duck, then it’s a duck.” For example, you can
verify that an object is iterable if it implemented the iterator protocol. For many
objects, this means it has a __iter__ “magic method,” though an alternative and better
way to check is to try using the iter function.

In [48]:
def isiterable(obj):
    try:
        iter(obj)
        return True
    except TypeError: # not iterable
        return False
    

This function would return True for strings as well as most Python collection types:

In [49]:
isiterable("a string")

True

In [50]:
isiterable([1, 2, 3])

True

In [51]:
isiterable(5)

False

A place where I use this functionality all the time is to write functions that can accept
multiple kinds of input. A common case is writing a function that can accept any
kind of sequence (list, tuple, ndarray) or even an iterator. You can first check if the
object is a list (or a NumPy array) and, if it is not, convert it to be one.

In [52]:
if not isinstance(x, list) and isiterable(x):
    x = list(x)

#### Imports

In Python a module is simply a file with the .py extension containing Python code.
Suppose that we had the following module.

In [54]:
# some_module.py
PI = 3.14159

def f(x):
    return x + 2


def g(a, b):
    return a + b

If we wanted to access the variables and functions defined in some_module.py, from
another file in the same directory we could do.

In [60]:
import some_module
result = some_module.f(5)
pi = some_module.PI

Or we could do this equivalently.

In [61]:
from some_module import f, g, PI

By using the as keyword you can give imports different variable names.

In [63]:
import some_module as sm
from some_module import PI as pi, g as gf

r1 = sm.f(pi)
r2 = gf(5, pi)

#### Binary operators and comparisons

Most of the binary math operations and comparisons are as you might expect.

In [64]:
5 - 7

-2

In [65]:
12 + 21.5

33.5

To check if two references refer to the same object, use the is keyword. is not is also
perfectly valid if you want to check that two objects are not the same.

In [66]:
a = [1, 2, 3]

b = a

c = list(a)

In [67]:
a is b

True

In [68]:
a is not c 

True

Since list always creates a new Python list (i.e., a copy), we can be sure that c is distinct
from a. Comparing with is is not the same as the == operator, because in this
case we have.

In [69]:
a == c

True

A very common use of is and is not is to check if a variable is None, since there is
only one instance of None.

In [70]:
a = None

In [71]:
a is None

True

All binary operators:

```
a + b Add a and b
a - b Subtract b from a
a * b Multiply a by b
a / b Divide a by b
a // b Floor-divide a by b, dropping any fractional remainder
a ** b Raise a to the b power
a & b True if both a and b are True; for integers, take the bitwise AND
a | b True if either a or b is True; for integers, take the bitwise OR
a ^ b For booleans, True if a or b is True, but not both; for integers, take the bitwise EXCLUSIVE-OR
a == b True if a equals b
a != b True if a is not equal to b
a <= b, a < b True if a is less than (less than or equal) to b
a > b, a >= b True if a is greater than (greater than or equal) to b
a is b True if a and b reference the same Python object
a is not b True if a and b reference different Python objects
```

#### Mutable and immutable objects

Most objects in Python, such as lists, dicts, NumPy arrays, and most user-defined
types (classes), are mutable. This means that the object or values that they contain can
be modified.

In [73]:
a_list = ["foo", 2, [4, 5]]
a_list[2] = (3, 4)
a_list

['foo', 2, (3, 4)]

Others, like strings and tuples, are immutable.

In [74]:
a_tuple = (3, 5, (4, 5))
a_tupe[1] = "four"

NameError: name 'a_tupe' is not defined

Remember that just because you can mutate an object does not mean that you always
should. Such actions are known as side effects. For example, when writing a function,
any side effects should be explicitly communicated to the user in the function’s documentation
or comments. If possible, I recommend trying to avoid side effects and
favor immutability, even though there may be mutable objects involved.

### Scalar types

Python along with its standard library has a small set of built-in types for handling
numerical data, strings, boolean (True or False) values, and dates and time. These
“single value” types are sometimes called scalar types and we refer to them in this
book as scalars. See Table 2-4 for a list of the main scalar types. Date and time handling
will be discussed separately, as these are provided by the datetime module in
the standard library.

```
- None The Python “null” value (only one instance of the None object exists)
- str String type; holds Unicode (UTF-8 encoded) strings
- bytes Raw ASCII bytes (or Unicode encoded as bytes)
- float Double-precision (64-bit) floating-point number (note there is no separate double type)
- bool A True or False value
- int Arbitrary precision signed integer
```

#### Strings

Python strings are immutable; you cannot modify a string.

In [75]:
a = "this is a string"
a[10] = "F"

TypeError: 'str' object does not support item assignment

Many Python objects can be converted to a string using the str function.

In [76]:
str(5.6)

'5.6'

Strings are a sequence of Unicode characters and therefore can be treated like other
sequences, such as lists and tuples.

In [78]:
s = "python"
list(s)

['p', 'y', 't', 'h', 'o', 'n']

And can be sliced.

In [79]:
s[3:]

'hon'

String templating or formatting is another important topic. The number of ways to
do so has expanded with the advent of Python 3, and here I will briefly describe the
mechanics of one of the main interfaces. String objects have a format method that
can be used to substitute formatted arguments into the string, producing a new
string.

In [80]:
template = "{0:.2f} {1:s} are worth US${2:d}"

In this string,
* {0:.2f} means to format the first argument as a floating-point number with two
decimal places.
* {1:s} means to format the second argument as a string.
* {2:d} means to format the third argument as an exact integer.

To substitute arguments for these format parameters, we pass a sequence of arguments
to the format method.

In [81]:
template.format(4.5560, "Argentine Pesos", 1)

'4.56 Argentine Pesos are worth US$1'

#### Bytes and unicode

In modern Python (i.e., Python 3.0 and up), Unicode has become the first-class string
type to enable more consistent handling of ASCII and non-ASCII text. In older versions
of Python, strings were all bytes without any explicit Unicode encoding. You
could convert to Unicode assuming you knew the character encoding. Let’s look at an
example.

In [82]:
val = "español"
val

'español'

We can convert this Unicode string to its UTF-8 bytes representation using the
encode method.

In [83]:
val_utf8 = val.encode("utf-8")
val_utf8

b'espa\xc3\xb1ol'

In [84]:
type(val_utf8)

bytes

Assuming you know the Unicode encoding of a bytes object, you can go back using
the decode method.

In [85]:
val_utf8.decode("utf-8")

'español'

While it’s become preferred to use UTF-8 for any encoding, for historical reasons you
may encounter data in any number of different encodings.

In [86]:
val.encode("latin1")

b'espa\xf1ol'

In [87]:
val.encode("utf-16")

b'\xff\xfee\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

In [88]:
val.encode("utf-16le")

b'e\x00s\x00p\x00a\x00\xf1\x00o\x00l\x00'

It is most common to encounter bytes objects in the context of working with files,
where implicitly decoding all data to Unicode strings may not be desired.

Though you may seldom need to do so, you can define your own byte literals by prefixing
a string with b.

In [90]:
bytes_val = b"this is bytes"
bytes_val

b'this is bytes'

In [92]:
decoded = bytes_val.decode("utf-8") # this is str (Unicode) now
decoded

'this is bytes'

[Strings documentation](https://docs.python.org/3.6/library/string.html)

#### None

None is the Python null value type. If a function does not explicitly return a value, it
implicitly returns None.

In [93]:
a = None

In [94]:
a is None

True

In [96]:
b = 5
b is not None

True

None is also a common default value for function arguments.

In [97]:
def add_and_maybe_multiply(a, b, c=None):
    result = a + b
    
    if c is not None:
        result = result * c
        
    return result

While a technical point, it’s worth bearing in mind that None is not only a reserved
keyword but also a unique instance of NoneType.

In [98]:
type(None)

NoneType

#### Dates and times

The built-in Python datetime module provides datetime, date, and time types. The
datetime type, as you may imagine, combines the information stored in date and
time and is the most commonly used.

In [99]:
from datetime import datetime, date, time

In [100]:
dt = datetime(2011, 10, 29, 20, 30, 21)

In [102]:
dt.day

29

In [103]:
dt.minute

30

Given a datetime instance, you can extract the equivalent date and time objects by
calling methods on the datetime of the same name.

In [104]:
dt.date()

datetime.date(2011, 10, 29)

In [105]:
dt.time()

datetime.time(20, 30, 21)

The strftime method formats a datetime as a string.

In [106]:
dt.strftime("%m/%d/%Y %H:%M")

'10/29/2011 20:30'

Strings can be converted (parsed) into datetime objects with the strptime function.

In [107]:
datetime.strptime("20091031", "%Y%m%d")

datetime.datetime(2009, 10, 31, 0, 0)

See Table 2-5 for a full list of format specifications.
```
%Y Four-digit year
%y Two-digit year
%m Two-digit month [01, 12]
%d Two-digit day [01, 31]
%H Hour (24-hour clock) [00, 23]
%I Hour (12-hour clock) [01, 12]
%M Two-digit minute [00, 59]
%S Second [00, 61] (seconds 60, 61 account for leap seconds)
%w Weekday as integer [0 (Sunday), 6]
%U Week number of the year [00, 53]; Sunday is considered the first day of the week, and days before the first Sunday of
the year are “week 0”
%W Week number of the year [00, 53]; Monday is considered the first day of the week, and days before the first Monday of
the year are “week 0”
%z UTC time zone offset as +HHMM or -HHMM; empty if time zone naive
%F Shortcut for %Y-%m-%d (e.g., 2012-4-18)
%D Shortcut for %m/%d/%y (e.g., 04/18/12)
```

When you are aggregating or otherwise grouping time series data, it will occasionally
be useful to replace time fields of a series of datetimes—for example, replacing the
minute and second fields with zero.

In [109]:
dt.replace(minute=0, second=0)

datetime.datetime(2011, 10, 29, 20, 0)

Since datetime.datetime is an immutable type, methods like these always produce
new objects.

The difference of two datetime objects produces a datetime.timedelta type.

In [110]:
dt2 = datetime(2011, 11, 15, 22, 30)

In [111]:
delta = dt2 - dt

In [112]:
delta

datetime.timedelta(17, 7179)

In [113]:
type(delta)

datetime.timedelta

The output timedelta(17, 7179) indicates that the timedelta encodes an offset of 17
days and 7,179 seconds.

Adding a timedelta to a datetime produces a new shifted datetime.

In [114]:
dt

datetime.datetime(2011, 10, 29, 20, 30, 21)

In [115]:
dt + delta

datetime.datetime(2011, 11, 15, 22, 30)

### Control flow

In [116]:
x = 2

if x < 0:
    pass
elif x == 0:
    pass
elif x > 0 or x > 2:
    pass
else:
    pass

pass is the “no-op” statement in Python.

In [119]:
for val in range(10):
    if val == 7:
        break
    if val == 3:
        continue
    print(val)

0
1
2
4
5
6


In [120]:
x = 256
total = 0
while x > 0:
    if total > 500:
        break
    total += x
    x = x // 2

The range function returns an iterator that yields a sequence of evenly spaced
integers.

In [121]:
range(10)

range(0, 10)

In [124]:
list(range(5, 0, -1))

[5, 4, 3, 2, 1]

As you can see, range produces integers up to but not including the endpoint.

ternary expression in Python allows you to combine an if-else block that produces
a value into a single line or expression. The syntax for this in Python is:

```
value = true-expr if condition else false-expr
```

Here, true-expr and false-expr can be any Python expressions. It has the identical
effect as the more verbose:
```
if condition:
value = true-expr
else:
value = false-expr
```

In [125]:
x = 5

In [126]:
"Non-negative" if x >= 0 else "Negative"

'Non-negative'

As with if-else blocks, only one of the expressions will be executed. Thus, the “if ”
and “else” sides of the ternary expression could contain costly computations, but only
the true branch is ever evaluated.

While it may be tempting to always use ternary expressions to condense your code,
realize that you may sacrifice readability if the condition as well as the true and false
expressions are very complex.